src/nunit.analyzers/ValuesUsage/ValuesUsageAnalyzer.cs

```diff
@@ -0,0 +1,91 @@
+                                                                                                                messageFormat: ValuesUsageAnalyzerConstants.ParameterTypeMismatchMessage,
+                                                                                                                category: Categories.Structure,
+                                                                                                                defaultSeverity: DiagnosticSeverity.Error,
+                                                                                                                description: ValuesUsageAnalyzerConstants.ParameterTypeMismatchDescription);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(parameterTypeMismatch);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private static void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var valuesType = context.Compilation.GetTypeByMetadataName(NUnitFrameworkConstants.FullNameOfTypeValuesAttribute);
+            if (valuesType is null)
+            {
+                return;
+            }
+
+            context.RegisterSymbolAction(symbolContext => AnalyzeParameter(symbolContext, valuesType), SymbolKind.Parameter);
+        }
+
+        private static void AnalyzeParameter(SymbolAnalysisContext symbolContext, INamedTypeSymbol valuesType)
+        {
+            var parameterSymbol = (IParameterSymbol)symbolContext.Symbol;
+            var attributes = parameterSymbol.GetAttributes();
+            if (attributes.Length == 0)
+            {
+                return;
+            }
+
+            foreach (var attribute in attributes.Where(x => x.ApplicationSyntaxReference is not null
+                                                       && SymbolEqualityComparer.Default.Equals(x.AttributeClass, valuesType)))
+            {
+                symbolContext.CancellationToken.ThrowIfCancellationRequested();
+                for (var index = 0; index < attribute.ConstructorArguments.Length; index++)
+                {
+                    var constructorArgument = attribute.ConstructorArguments[index];
+
+                    // If the compiler detects an illegal constant, we shouldn't check it.
+                    // Unfortunately the constant 'null' is marked as Error with a null type.
+                    if (constructorArgument.Kind == TypedConstantKind.Error && constructorArgument.Type is not null)
+                    {
+                        continue;
```

Comment:
We should still check whether `null` is a legal argument for the parameter.
It would be legal if the parameter is not a ValueType (struct or int etc).
I can't recall if we can detect nullability, in which case `null` would only be valid if the parameter is marked nullable.

Reply:
Huh, I just tried out
```csharp
[Test]
public void Blah([Values(null, 0)] int numberValue)
{
    Assert.That(numberValue, Is.EqualTo(0));
}
```
which passed.  So the behaviour for null is a little weird (seems to end up being `default(T)`).  I can't imagine we want to encourage this behaviour, but equally the analyzer claiming something will fail when it will run correctly isn't particularly helpful either.

This means that my current test `AnalyzeWhenArgumentTypeIsIncorrect` actually warns incorrectly, since that test would run without issues.  Amusingly enough, the TestCaseUsageAnalyzer has the same bug, in that null ends up being converted to `default(T)` so the test will actually run

Reply:
>  I can't imagine we want to encourage this behaviour, but equally the analyzer claiming something will fail when it will run correctly isn't particularly helpful either.

Analyzers not only warn against failing code, but are also for improving. You could create a separate diagnostic saying that `null` should be `default(T)`.

Reply:
For the general point, `CanAssignTo` does handle null conversions correctly (including nullable structs), so the logic is correct as far as I can tell.  (We do not check for nullability annotations in that method.)

Reply:
We will leave that for another separate issue.